addons/hr_gamification/models/gamification.py

```diff
@@ -1,45 +1,35 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
+from openerp import _, api, fields, models
+from openerp.exceptions import ValidationError
 
 
-class hr_gamification_badge_user(osv.Model):
+class GamificationBadgeUser(models.Model):
     """User having received a badge"""
 
     _name = 'gamification.badge.user'
     _inherit = ['gamification.badge.user']
 
-    _columns = {
-        'employee_id': fields.many2one("hr.employee", string='Employee'),
-    }
+    employee_id = fields.Many2one('hr.employee', string='Employee')
 
-    def _check_employee_related_user(self, cr, uid, ids, context=None):
-        for badge_user in self.browse(cr, uid, ids, context=context):
-            if badge_user.user_id and badge_user.employee_id:
-                if badge_user.employee_id not in badge_user.user_id.employee_ids:
-                    return False
-        return True
+    @api.one
+    @api.constrains('employee_id')
+    def _check_employee_id(self):
+        if self.user_id and self.employee_id and self.employee_id not in self.user_id.employee_ids:
+            raise ValidationError(_('The selected employee does not correspond to the selected user.'))
 
-    _constraints = [
-        (_check_employee_related_user, "The selected employee does not correspond to the selected user.", ['employee_id']),
-    ]
 
-
-class gamification_badge(osv.Model):
+class GamificationBadge(models.Model):
     _name = 'gamification.badge'
     _inherit = ['gamification.badge']
 
-    def get_granted_employees(self, cr, uid, badge_ids, context=None):
-        if context is None:
-            context = {}
-
-        employee_ids = []
-        badge_user_ids = self.pool.get('gamification.badge.user').search(cr, uid, [('badge_id', 'in', badge_ids), ('employee_id', '!=', False)], context=context)
-        for badge_user in self.pool.get('gamification.badge.user').browse(cr, uid, badge_user_ids, context):
-            employee_ids.append(badge_user.employee_id.id)
-        # remove duplicates
-        employee_ids = list(set(employee_ids))
+    @api.multi
+    def get_granted_employees(self):
+        employee_ids = self.env['gamification.badge.user'].search([
+            ('badge_id', 'in', self.ids),
+            ('employee_id', '!=', False)
+        ]).mapped('employee_id.id')
```

Comment:
Might be interesting to see how `self.mapped('owner_ids').filtered('employee_id').mapped('employee_id.id')` compares, in theory it should return the same value but I don't know how efficient it'd be.